packages/@aws-cdk/aws-redshift/lib/database-query-provider/create-table.ts

```diff
@@ -0,0 +1,71 @@
+/* eslint-disable-next-line import/no-unresolved */
+import * as AWSLambda from 'aws-lambda';
+import { ClusterProps, executeStatement, getClusterPropsFromEvent, getResourceNameFromPhysicalId, getResourceProperty, makePhysicalId } from './util';
+
+interface Column {
+  name: string;
+  dataType: string;
+}
+
+export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
+  const tableColumns = JSON.parse(getResourceProperty('tableColumns', event.ResourceProperties)) as Column[];
+  const tableName = event.ResourceProperties.tableName;
+  const clusterProps = getClusterPropsFromEvent(event.ResourceProperties);
+
+  if (event.RequestType === 'Create') {
+    await createTable(tableName, tableColumns, clusterProps);
+    return { PhysicalResourceId: makePhysicalId(tableName, clusterProps), Data: { tableName: tableName } };
+  } else if (event.RequestType === 'Delete') {
+    await dropTable(getResourceNameFromPhysicalId(event.PhysicalResourceId), clusterProps);
+    return;
+  } else if (event.RequestType === 'Update') {
+    await updateTable(tableName, tableColumns, clusterProps, event.OldResourceProperties);
+    return { PhysicalResourceId: makePhysicalId(tableName, clusterProps), Data: { tableName: tableName } };
+  } else {
+    /* eslint-disable-next-line dot-notation */
+    throw new Error(`Unrecognized event type: ${event['RequestType']}`);
+  }
+}
+
+async function createTable(tableName: string, tableColumns: Column[], clusterProps: ClusterProps) {
+  const tableColumnsString = tableColumns.map(column => `${column.name} ${column.dataType}`).join();
+  await executeStatement(`CREATE TABLE ${tableName} (${tableColumnsString})`, clusterProps);
+}
+
+async function dropTable(tableName: string, clusterProps: ClusterProps) {
+  await executeStatement(`DROP TABLE ${tableName}`, clusterProps);
+}
+
+async function updateTable(tableName: string, tableColumns: Column[], clusterProps: ClusterProps, oldResourceProperties: { [Key: string]: any }) {
+  const oldClusterProps = getClusterPropsFromEvent(oldResourceProperties);
+  if (clusterProps !== oldClusterProps) {
+    return createTable(tableName, tableColumns, clusterProps);
+  }
+
+  const oldTableColumns = JSON.parse(getResourceProperty('tableColumns', oldResourceProperties)) as Column[];
+  const oldTableName = oldResourceProperties.tableName;
+
+  if (tableName !== oldTableName) {
+    return createTable(tableName, tableColumns, clusterProps);
+  }
+
+  const changes: string[] = [];
+  tableColumns.forEach(tableColumn => {
+    const oldTableColumn = oldTableColumns.find(({ name: oldName }) => tableColumn.name === oldName);
+    if (!oldTableColumn) {
+      changes.push(`ADD ${tableColumn.name} ${tableColumn.dataType}`);
```

Comment:
Couldn't this result in changing the order of the columns? Actually, the order not matching the order that they are listed in the `Table` construct definition. I am thinking for the s3 copy command for firehose, the list of columns need to be in the same order as they actually are in the table. I wonder if this will cause some weird edge cases.

Reply:
I think you're referring to the following Redshift documentation ([source](https://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-column-mapping.html)):

> You can specify a comma-separated list of column names to load source data fields into specific target columns. The columns can be in any order in the COPY statement, but when loading from flat files, such as in an Amazon S3 bucket, their order must match the order of the source data. 

The way that I read this is: the list of columns in the COPY command and the S3 objects must be in the same order, but the list of columns in the COPY command will be matched to the columns in the Redshift table by name (in any order).

I will try to confirm this

Reply:
Confirmed!